File: Raven.Client.Java/lib/src/main/java/raven/abstractions/oauth/AbstractAuthenticator.java
package raven.abstractions.oauth;

import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpRequest;
import org.apache.http.client.methods.HttpUriRequest;

import raven.abstractions.closure.Action1;
import raven.abstractions.connection.WebRequestEventArgs;

public abstract class AbstractAuthenticator {

  protected String currentOauthToken;

  public void configureRequest(Object sender, WebRequestEventArgs e) {
    if (StringUtils.isEmpty(currentOauthToken)) {
      return;
    }
    setHeader(e.getRequest(), "Authorization", currentOauthToken);
  }

  protected static void setHeader(HttpUriRequest request, String key, String value) {
    request.setHeader(key, value);
  }

  public abstract Action1<HttpRequest> doOAuthRequest(String oauthSource);

}

File: src/Raven.Client/OAuth/AbstractAuthenticator.cs
using System;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using Raven.Abstractions.Connection;

namespace Raven.Abstractions.OAuth
{
    public abstract class AbstractAuthenticator
    {
        public string CurrentToken { get; set; }
        public string CurrentTokenWithBearer { get; set; }

        public virtual void ConfigureRequest(object sender, WebRequestEventArgs e)
        {
            SetAuthorization(e);
        }

        protected void SetAuthorization(WebRequestEventArgs e)
        {
            if (string.IsNullOrEmpty(CurrentToken))
                return;

            if (e.Client == null)
                return;
            SetAuthorization(e.Client);
        }

        protected void SetAuthorization(HttpClient e)
        {
            if (string.IsNullOrEmpty(CurrentToken))
                return;

            try
            {
                e.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", CurrentToken);
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException(string.Format("Could not set the Authorization to the value 'Bearer {0}'", CurrentToken), ex);
            }
        }

        protected static void SetHeader(WebHeaderCollection headers, string key, string value)
        {
            try
            {
                headers[key] = value;
            }
            catch (Exception e)
            {
                throw new InvalidOperationException("Could not set '" + key + "' = '" + value + "'", e);
            }
        }
    }
}
